### ISSUE TITLE
Solve one-to-many problem with genetic markers that map to multiple genomes

### ISSUE THREAD

**Issue by sammyjava** [legumeinfo/mine-issues#26]
A really core-level change would be to have a non-SequenceFeature GeneticMarker class and then a SequenceFeature to which it is related (via collection) called, say, GenomicMarker. Then we can load any genomes and have the GeneticMarker "map" to those genomes via the related "GenomicMarker." That adds an extra step, but if we want multiple genomes in the mines with markers mapped to them, we pretty much have to do this. (The collections would be populated by a post-processor, since you'd have to load all the "GenomicMarkers" to make the collections. That's normal.)

_Originally posted by @sammyjava in https://github.com/legumeinfo/datastore-issues/issues/17#issuecomment-856209023_

**Comment by sammyjava** [comment]
I've been avoiding this discussion until the files got nailed down. Since I need to merge markers from genetic files (by secondaryIdentifier) with markers in genomic files (full yuck primaryIdentifier) and we may have those markers mapped to multiple genomes, there probably is no way around separating GeneticMarker from GenomicMarker in the mines.

But that's a loader issue now, the files are all set. The question is just when I make that change and rebuild -- now, or after I get all the files sorted for all the genus mines and, finally, LegumeMine. Since file curation is a bigger job, I've been focusing on that.

**Comment by sammyjava** [comment]
@cann0010 I'm about to implement this so I can load multiple genomes into SoyMine. As I recall you're on board with it, but here's a heads up that it's becoming a real thing. GeneticMarker will become a BioEntity, not a sequence feature, which contains a collection genomicMarkers which are SequenceFeatures placed on genomes for which there are marker GFFs.

**Comment by sammyjava** [comment]
This gets a bit funky for GWAS displays, which show traits and markers on a specific genome. We may need a definition of "current" or "default" genome for that, which is something we've talked about in the past. Or I can just hold off on this and load a single Wm82 genome, etc., for now.

**Comment by StevenCannon-USDA** [comment]
Thanks for the heads-up/reminder.
Can you point to one file as an example?

**Comment by sammyjava** [comment]
There's no file, this is strictly about the mine data model. Same files as always. But because we load markers from genetic files (where they simply have a name) and we load marker positions on genomes from GFFs, we have to relate a genetic-file-originated marker to a genomic-file-originated marker. InterMine does not support a SequenceFeature residing on more than one genome. So to support placing a marker on multiple genomes I need to have a bridge, GenomicMarker, which connects a GeneticMarker to a particular genome assembly. This will, of course, affect folks writing imjs queries on markers, so it's probably better done sooner than later.

**Comment by StevenCannon-USDA** [comment]
OK, got it. Change away ...

**Comment by sammyjava** [comment]
And, actually, GeneticMarker shouldn't even be a BioEntity, because that has a specific strain. It's a step above that, which is simply Annotatable (something that's related to a publication and/or ontology term). Markers are strain-independent diagnostic quantities.

**Comment by StevenCannon-USDA** [comment]
Agree.

**Comment by sammyjava** [comment]
I never got around to this and will not be doing data model and loader updates at this point. So this is something for @svengato and @adf-ncgr to ponder in the future. It's a fairly major update.

**Comment by sammyjava** [comment]
I kinda forgot about this and blithely went ahead and started loading genome-mapped GeneticMarker objects from GFFs. Before I've loaded the markers associated with maps, etc., in the /genetic/ collections. I'll ponder this some more, not sure if I'll come up with anything different. If not, I'll have to create the GenomicMarker class and load those again. But they're called genetic_marker in the GFFs and I'd like to figure out some way of doing this. I could just have GeneticMarker contain a collection of GeneticMarkers (relatedMarkers? similarMarkers?) and load the collection in a post-processor like I do with a lot of other related stuff.

**Comment by sammyjava** [comment]
Now I'm thinking of loading the marker from the /genetic/ collections simply as a string identifier (markerName) with a collection (geneticMarkers) left unfilled by the DS loader. Then, in post-processing, objects will get their geneticMarkers collections filled by all the markers that match markerName. Here's a concrete example:
```
GWASResult
identifier=Seed weight 5-g1
trait=Seed weight
markerName=ss715587065
pValue=7.27E-05
geneticMarkers={glyma.Wm82.gnm1.ss715587065,glyma.Wm82.gnm2.ss715587065}
```
This gets the job done in that the GWASResult report (for this example) will list the genome-located markers while simply listing the name of the marker in the header. It's a bit trickier with QTL-marker and LinkageGroup-Marker relationships since that's one-to-many distinct markers (one LG/QTL has several associated markerNames) but I can work that out by storing a delimited string ("ss715639553|ss715639323") as markerNames.

This avoids creating a separate marker class (GenomicMarker) which is more of a kludge IMO.

**Comment by sammyjava** [comment]
I'm liking my solution to this, which is now implemented in 5.0.6.1 on dev (I've added an additional number to the IM version now to avoid the repeated versionitis problems I've had with restarting production mines, among other reasons). However, a buggery thing that happens is that the GWAS displayer, which shows a Manhattan plot on a particular genome assembly, now doesn't know which genome assembly to use: the markers are mapped to various genome versions. So I think I'll have to add a selector to that displayer to choose which genome version you want to view (with a default). Which is a major update, but I think it's needed so we can keep genetic data assembly-agnostic.
